### PR TITLE
Ruckus ICX Alpha Phase 1

### PR DESCRIPTION
**Pull request by Static0verride**

### CODE REVIEW

**Comment by itdependsnetworks** [comment]
Just a few notes:
- I think many of these methods could be set as private methods. e.g. `find_words` -> `_ find_words`
- The traditional naming of `napalm-ruckus-fastiron/fi.py` is `napalm_ruckus_fastiron/ruckus_fastiron.py`. It probably makes sense to say the same. 
- You may be able to make use of: https://github.com/napalm-automation/napalm/blob/develop/napalm/base/utils/string_parsers.py#L79 for date time parsing.
- With getters, it often makes sense to either know the data or fail. Meaning, if you split the data, it should be X columns, if it is not, you got something you were not expecting. The else should be a failure, indicating you had an unknown parsing issue on line `current_line`.
e.g.
```
        for val in output:
            val = val.split()
            arp_counters.update({
                'interfaces': val[5],
                'mac': val[2],
                'ip': val[1],
                'age': val[4],
            })
```
becomes
```
        for val in output:
            _, ip, mac, _, age, interface = val.split()
            arp_counters.update({
                'interfaces': interface,
                'mac': mac,
                'ip': ip,
                'age': age,
            })
```
and 
```
        for val in shw_int_speed:                               # speed words contained and compared
            if val == 'auto,' or val == '1Gbit,':               # appends speed hat
                speed.append(1000)
<omitted>
            else:
                speed.append(100000)
```
becomes
```
        for val in shw_int_speed:                               # speed words contained and compared
            if val == 'auto,' or val == '1Gbit,':               # appends speed hat
                speed.append(1000)
<omitted>
            elif val == '100Gbit':
                speed.append(100000)
            else:
                Raise MyCustomException('Unknown Error')
```

File: napalm-ruckus-fastiron/fi.py [hunk 1]
```diff
@@ -0,0 +1,1378 @@
...
+        if len(word_list) == 0 or len(pos_list) == 0:   # returns NONE if list is empty
+            return None
+
+        size = len(word_list)
+        sentence = output.split()                   # breaks long string into separate strings
+
+        for m in range(0, size):                    # Iterates through size of word list
+            pos = int(pos_list.pop())               # pops element position and word pair in list
+            word = word_list.pop()
+            if sentence.__contains__(word):         # checks if word is contained in text
```
Comment (bewing) [inline napalm-ruckus-fastiron/fi.py]: Think all the `__contains__` calls might be more pythonic/readable if rewritten as `if word in sentence:`

Comment (bewing) [reply napalm-ruckus-fastiron/fi.py]: Maybe not for this PR, but need to start thinking about the test framework -- you can look at https://github.com/napalm-automation-community/napalm-mos/tree/develop/test/unit for a barebones example

**Comment by bewing** [comment]
Maybe not necessary for this PR, but need to start thinking about unit tests -- can look at https://github.com/napalm-automation-community/napalm-mos/tree/develop/test/unit for a barebones example.

File: napalm-ruckus-fastiron/fi.py [hunk 2]
```diff
@@ -0,0 +1,1378 @@
...
+        temp_list = list()                          # Creates a list to store variables
+        for a_string in nline_list:                 # iterates through list
+            if a_string .__contains__(del_word):    # if word matches, word is skipped
+                continue
+            else:
+                temp_list.append(a_string.split())  # Word didn't match store in list
+        return temp_list
+
+    @staticmethod
+    def facts_uptime(my_string):  # TODO check for hours its missing....
```
Comment (bewing) [inline napalm-ruckus-fastiron/fi.py]: Perhaps possible to just call `napalm.ios.IOSDriver.parse_uptime` ?
https://github.com/napalm-automation/napalm/blob/develop/napalm/ios/ios.py#L901

File: napalm-ruckus-fastiron/fi.py [hunk 3]
```diff
@@ -0,0 +1,1378 @@
...
+        interfaces_list = list()
+        n_line_output = FastIronDriver.creates_list_of_nlines(shw_int_brief)
+        interface_details = FastIronDriver.delete_if_contains(n_line_output, del_word)
+
+        for port_det in interface_details:
+
+            if trigger == 0:
+                interfaces_list.append(port_det[pos])
+            else:
+                if port_det[pos].__contains__("ve") or port_det[pos].__contains__("lb") or \
```
Comment (bewing) [inline napalm-ruckus-fastiron/fi.py]:
```python
if any(x in port_det[pos] for x in ["ve", "lb", "tunnel"):
     continue
```

File: napalm-ruckus-fastiron/fi.py (hunk 1 above)
Comment (Static0verride) [reply napalm-ruckus-fastiron/fi.py]: Great point, I agree much more readable. Will make this change.

File: napalm-ruckus-fastiron/fi.py (hunk 3 above)
Comment (Static0verride) [reply napalm-ruckus-fastiron/fi.py]: Thank you to introducing me to any, this has been changed :)

File: napalm-ruckus-fastiron/ruckus_fastiron.py [hunk 4]
```diff
@@ -0,0 +1,1405 @@
...
+        vrf_interface = dict()
+        check = self.device.send_command('show version')
+
+        if check.__contains__('7150'):                              # ICX7150 does not support VRF
+            return {}
+
+        if check.__contains__('SPR'):                               # check if router version
+            vrf_type = 'L3VRF'                                      # adds vrf of type router
+        else:
+            vrf_type = 'L2VRF'                                      # adds vrf of type switch
```
Comment (bewing) [inline napalm-ruckus-fastiron/ruckus_fastiron.py]: When IETF/openconfig is specifying L3VRF/L2VRF, they're specifically referring to L3VPNs and L2VPNs (EVPN, psuedowire, etc), not the model/type of device containing the VRF.  I imagine all the VRFs on this device would be the L3VRF type.

File: napalm-ruckus-fastiron/fi.py (hunk 2 above)
Comment (Static0verride) [reply napalm-ruckus-fastiron/fi.py]: I looked into there implementation and this wont be possible without some changes in there function.
Reason:
they  ".split()" the string only after a "," is found and assume that the first value after the comma (split) will be a numerical value.  This is due to the ways there "show version" data is structured.  We do not have the same structure :(

My current implementation finds the time keywords (seconds, minutes, hours, days...) and grabs the value before it as they are pre-fix to the keywords.

I will revise this portion and make the code more readable :). This reminds me i should look at not reinventing the wheel. Thank you for the great suggestion!

File: napalm-ruckus-fastiron/ruckus_fastiron.py (hunk 4 above)
Comment (Static0verride) [reply napalm-ruckus-fastiron/ruckus_fastiron.py]: Thanks for the tip. Ruckus ICX is unique in that not all models support L3/L2VRF or any device running on switch image. I have added the following code that will return an empty dictionary if VRF is not supported. Currently working on cases for VRF configured 

`        if any(x in check for x in ["7150", "SPS"]):                # ICX7150 does not support VRF
            return {}                                               # neither does switch image

        if "7250" in check:
            cur_version = FastIronDriver.__retrieve_all_locations(check, 'Version', 1)

            if cur_version.pop() > "8.0.50":
                pass
            else:
                return {}`

**Comment by Static0verride** [comment]
Hello Napalm Community, what steps are additionally required so i may take ownership of this repo? Thank you and kind regards

**Comment by mirceaulinic** [comment]
Hey @Static0verride are you still working on this, think we can get this in and release the first version of this driver anytime soon?

**Comment by mirceaulinic** [comment]
You are right @Static0verride - thanks for this addition. I have added you as a maintainer of this repository, let me know if you need anything from me. :-)

**Pull request merged**